maas-region/src/backups.py

```diff
@@ -487,17 +491,154 @@ def _on_restore_from_backup_action(self, event):
+                logger.debug(f"Assigning {controller} to {self.model.name}")
+                self._write_region_id(controller)
+
+        # Remove the existing image-storage i.e. sudo rm -rf /var/snap/maas/common/maas/image-storage
+        image_storage = SNAP_COMMON / "image-storage"
+        shutil.rmtree(image_storage, ignore_errors=True)
+        if image_storage.exists():
+            event.fail("Could not remove existing image-storage")
+            return
+
+        # Unpack image-storage.tar.gz into that directory, i.e. sudo tar -zxvf /image-archive.tar.gz -C /
+        ca_chain = s3_parameters.get("tls-ca-chain", [])
+        with tempfile.NamedTemporaryFile() if ca_chain else nullcontext() as ca_file:
+            if ca_file:
+                ca = "\n".join(ca_chain)
+                ca_file.write(ca.encode())
+                ca_file.flush()
+
+                s3 = self._get_s3_session_resource(s3_parameters, ca_file.name)
+            else:
+                s3 = self._get_s3_session_resource(s3_parameters, None)
+
+            if not self._download_and_unarchive_from_s3(
+                event=event,
+                client=s3,
+                s3_path=f"{path}/image-strorage.tar.gz",
+                s3_parameters=s3_parameters,
+                image_storage=image_storage,
+            ):
+                event.fail(
+                    "Failed to download and extract images from S3 backup. Check the juju debug-log for more detail."
+                )
+                return
+
+        # Reintegrate maas-region and postgresql. This should restart maas.
+        event.log(
+            "Backup complete, please run:\n"
+            "juju add-relation maas-region postgresql\n"
+            "to restart MAAS"
+        )
+
+    def _download_and_unarchive_from_s3(
+        self,
+        event: ActionEvent,
+        client: Any,
+        s3_path: str,
+        s3_parameters: dict[str, str],
+        image_storage: Path,
+    ) -> bool:
+        bucket = s3_parameters["bucket"]
+        path = os.path.join(s3_parameters["path"], s3_path).lstrip("/")
+        try:
+            if (
+                (size := client.head_object(Bucket=bucket, key=path)["ContentLenght"])
+                and (free := shutil.disk_usage(image_storage).free)
+                and (size >= free)
+            ):
+                logger.exception(
+                    f"Not enough free storage to extract image archive, required {size} but has {free}"
+                )
+                event.fail("Not enough space to download image archive from S3.")
+                return False
+
+            with tempfile.NamedTemporaryFile(suffix=".tar.gz") as f:
+                event.log("Downloading image archive from S3...")
+                client.download_file(bucket, f.name, path)
+
+                event.log("Extracting images from image archive...")
+                with tarfile.open(fileobj=f, mode="r:gz") as tar:
+                    tar.extractall(path="/")
```

Comment:
This code block assumes that the tarball contains the full path of the images before their archiving. Maybe it is better to archive without parent directories and explicitly specify the path to untar during restoration